### ISSUE TITLE
[$1000 Bounty] Training workload for CNN PPO on ALE/Breakout-v5

### ISSUE THREAD

**Issue by vladimirjovanovicTT** [tenstorrent/tt-blacksmith#454]
## Add CNN PPO Single-Chip RL Training on ALE/Breakout-v5 on TT-N150 with CPU Baseline  

**Background:**  
- TT-Blacksmith is a growing collection of model demos showcasing the capabilities of AI models running on Tenstorrent hardware.  
- This bounty aims to bring up **full training** of a **CNN-based PPO** (Proximal Policy Optimization) reinforcement learning agent on the **ALE/Breakout-v5** Atari environment on a single **Tenstorrent N150** chip, faithfully reproducing the training setup in PyTorch.  
- To establish a baseline, the same training must also be run on CPU and the results compared, ensuring that metrics (e.g., episode reward curves, policy loss, value loss) align between CPU and TT-N150 runs.  

When encountering compilation or runtime issues, contributors are expected to:  
1. Delegate the issue with a minimal repro to the appropriate Tenstorrent repository (`tt-xla/tt-mlir` for compilation, `tt-metal` for runtime).  
2. Provide a **CPU fallback workaround** in their training code so training can continue while the issue is investigated.  

This bounty is an opportunity for AI developers to:  
- Contribute to Tenstorrent's model demos  
- Validate single-chip reinforcement learning training  
- Earn rewards  

---

**Requirements:**  
- **Framework:** Implementation must be in **PyTorch**.  

1. **CNN PPO Training on TT-N150:**  
   - Implement **end-to-end PPO training** with a CNN policy network on a single chip.  
   - Train on the **ALE/Breakout-v5** environment.  
   - Reproduce all training hyperparameters, environment setup, and workflow from standard PPO Atari implementations.  

2. **CPU Baseline:**  
   - Run the same training on CPU.  
   - Compare results, ensuring parity in metrics and convergence.  

3. **Fallback Mechanism:**  
   - In cases where an operation fails on TT-N150, redirect only the **minimal possible function containing that operation** to CPU.  
   - Do **not** fallback the entire training step to CPU — training must primarily execute on TT hardware.  

4. **Verification:**  
   - To confirm training is running on TT hardware, set:  
     ```bash
     export TTXLA_LOGGER_LEVEL=DEBUG
     ```  
     - Look for printed TTIR graphs (e.g., `graph module @jit_training_step`) confirming TT execution.  

5. **Koyeb Machines:**  
   - Training should be run on **Koyeb-hosted N150 instances**.  
   - Request access here: [https://www.koyeb.com/solutions/tenstorrent](https://www.koyeb.com/solutions/tenstorrent)  
   - Wait for the onboarding email from Koyeb with setup instructions.  

6. **Documentation:**  
   - Provide a detailed **README.md** including:  
     - Setup instructions for both CPU and TT-N150  
     - Explanation of PPO setup and environment configuration  
     - Explanation of fallback logic  

7. **Examples:**  
   - Include sample output and **plots of episode reward curves** for CPU vs TT-N150 runs.  

8. **Dependencies:**  
   - Clearly document installation steps and dependencies.  

---

**Contribution Guidelines:**  
1. Fork the **tt-blacksmith** repository.  
2. Create a directory:  
   ```
   tt-blacksmith/blacksmith/experiments/torch/BOUNTIES/ppo-breakout
   ```  
3. Add your model implementation, training scripts, and documentation there.  
4. Submit **2 pull requests**:  
   - **PR 1 — CPU baseline:** A working training example on CPU that demonstrates correct training (loss going down, validation metrics, etc.).  
   - **PR 2 — TT device execution:** Move training to TT device execution, including fallback logic and CPU vs TT-N150 comparison.  
5. Each PR should include:  
   - A detailed description  
   - Results and comparison plots  
   - Relevant information to help reviewers evaluate your contribution  

---

**Evaluation Criteria:**  
1. **Framework Compliance:** Implementation must be in PyTorch.  
2. **Hardware Requirement:** Must demonstrate training on a single TT-N150 chip (via Koyeb).  
3. **Completeness:** End-to-end PPO training on ALE/Breakout-v5 must be functional.  
4. **Metric Parity:** CPU vs TT-N150 runs should show comparable convergence (reward curves, losses, etc.).  
5. **Fallback Implementation:** Only targeted CPU fallbacks are acceptable — the full training step must not run on CPU.  
6. **Verification:** Use `TTXLA_LOGGER_LEVEL=DEBUG` to confirm TTIR execution of key training functions.  
7. **Minimal Repros:** Any delegated issue must include a minimal reproducer.  
8. **Code Quality:** Clear, maintainable, and well-commented code.  
9. **Documentation:** Setup and usage must be clearly explained, including environment setup.

**Comment by jberkowitzTT** [comment]
Hi @Negoje, thanks for the great work so far! Based on Contribution Guidelines, there should be 2 PRs. 1 for the CPU & one for TT device execution.

Are you working on the second half / do you have timelines in mind for that?
